R/appMetadata.R

```diff
@@ -0,0 +1,378 @@
+    documentsHavePython = documentsHavePython,
+    quartoInfo = quartoInfo
+  )
+}
+
+# infer the mode of the application from its layout
+# unless we're an API, in which case, we're API mode.
+inferAppMode <- function(appDir,
+                         appPrimaryDoc = NULL,
+                         appFiles = NULL,
+                         quartoInfo = NULL,
+                         isCloudServer = FALSE) {
+
+  if (is.null(appFiles)) {
+    appFiles <- bundleFiles(appDir)
+  }
+
+  # TODO(HW): detection rules only apply to files in top-level directory
+  # Should make that explicit in the code
+
+  # plumber API
+  plumberFiles <- grep("^(plumber|entrypoint).r$", appFiles, ignore.case = TRUE, perl = TRUE)
+  if (length(plumberFiles) > 0) {
+    return("api")
+  }
+
+  # TODO(HW): I think this special case should live in appMetadata so that
+  # inferAppMode no longer needs the `appPrimaryDoc` argument
+  if (!is.null(appPrimaryDoc) &&
+      tolower(tools::file_ext(appPrimaryDoc)) == "r") {
+    return("shiny")
+  }
+
+  # Shiny application using single-file app.R style.
+  appR <- grep("^app.r$", appFiles, ignore.case = TRUE, perl = TRUE)
+  if (length(appR) > 0) {
+    return("shiny")
+  }
+
+  # Determine if we have Rmd files, and if they use the Shiny runtime.
+  rmdFiles <- grep("^[^/\\\\]+\\.rmd$", appFiles, ignore.case = TRUE, perl = TRUE, value = TRUE)
+  shinyRmdFiles <- sapply(file.path(appDir, rmdFiles), isShinyRmd)
+
+  # Determine if we have qmd files, and if they use the Shiny runtime
+  qmdFiles <- grep("^[^/\\\\]+\\.qmd$", appFiles, ignore.case = TRUE, perl = TRUE, value = TRUE)
+  shinyQmdFiles <- sapply(file.path(appDir, qmdFiles), isShinyRmd)
+
+  # We make Quarto requirement conditional on the presence of files that Quarto
+  # can render and _quarto.yml, because keying off the presence of qmds
+  # *or* _quarto.yml was causing deployment failures in static content.
+  # https://github.com/rstudio/rstudio/issues/11444
+  hasQuartoYaml <- any(grepl("^_quarto.y(a)?ml$", x = appFiles, ignore.case = TRUE, perl = TRUE))
+  hasQuartoCompatibleFiles <- length(qmdFiles) > 0 || length(rmdFiles > 0)
+  requiresQuarto <- (hasQuartoCompatibleFiles && hasQuartoYaml) || length(qmdFiles) > 0
+
+  # We gate the deployment of content that appears to be Quarto behind the
+  # presence of Quarto metadata. Rmd files can still be deployed as Quarto
+  # content.
+  # TODO(HW): I think this logic should be moved into appMetdata() because this
+  # should also error if the user specified appMode = "quarto-shiny"
+  if (requiresQuarto && is.null(quartoInfo)) {
+    stop(paste(
+      "Attempting to deploy Quarto content without Quarto metadata.",
+      "Please provide the path to a quarto binary to the 'quarto' argument."
+    ))
+  }
+
+  # Shiny or Quarto documents with "server: shiny" in their YAML front matter
+  # are rmd-shiny or quarto-shiny.
+  if (any(shinyRmdFiles) || any(shinyQmdFiles)) {
+    if (!is.null(quartoInfo)) {
```

Comment:
@toph-allen can you comment on why we use mode `quarto-shiny` when you have Rmd files and `quarto` set?

Reply:
Don't worry, found #595